In [2]:
from bs4 import BeautifulSoup #library used to scrape
import requests               #library used to actually get the results


# Web Scraping of Google Scholar Articles

In the URL, The keyword: Industrial Control System.
Start year (as_ylo): 2000
End year(as_yhi):2020
number of articles:all published in this year range

In [3]:
headers = {'user-agent':'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36'} #sring which identifies crawler as a web browser
url='https://scholar.google.com/scholar?q=Industrial+Control+System&hl=en&as_sdt=0%2C5&as_ylo=2000&as_yhi=2020'

In [4]:
response=requests.get(url,headers=headers)

In [5]:
soup=BeautifulSoup(response.content,'lxml')#conversion into xml document

In [6]:
for item in soup.select('[data-lid]'):
  print(item)
  print(item.select('h3')[0].get_text())
  item.select('a')[0]['href']
  print('----------')

<div class="gs_r gs_or gs_scl" data-aid="kvjW9FoAkBQJ" data-cid="kvjW9FoAkBQJ" data-did="kvjW9FoAkBQJ" data-lid="" data-rp="0"><div class="gs_ggs gs_fl"><div class="gs_ggsd"><div class="gs_or_ggsm" ontouchstart="gs_evt_dsp(event)" tabindex="-1"><a data-clk="hl=en&amp;sa=T&amp;oi=gga&amp;ct=gga&amp;cd=0&amp;d=1481684668059678866&amp;ei=jJN4ZYuzCpHcyQS51pfoCA" data-clk-atid="kvjW9FoAkBQJ" href="http://www.gocs.com.de/pages/fachberichte/archiv/164-sp800_82_r2_draft.pdf"><span class="gs_ctg2">[PDF]</span> gocs.com.de</a></div></div></div><div class="gs_ri"><h3 class="gs_rt" ontouchstart="gs_evt_dsp(event)"><span class="gs_ctc"><span class="gs_ct1">[PDF]</span><span class="gs_ct2">[PDF]</span></span> <a data-clk="hl=en&amp;sa=T&amp;oi=ggp&amp;ct=res&amp;cd=0&amp;d=1481684668059678866&amp;ei=jJN4ZYuzCpHcyQS51pfoCA" data-clk-atid="kvjW9FoAkBQJ" href="http://www.gocs.com.de/pages/fachberichte/archiv/164-sp800_82_r2_draft.pdf" id="kvjW9FoAkBQJ">Guide to <b>industrial control systems </b>(ICS) s

In [7]:
for item in soup.select('[data-lid]'):
   print(item.select('a')[0]['href'])

http://www.gocs.com.de/pages/fachberichte/archiv/164-sp800_82_r2_draft.pdf
https://www.scienceopen.com/document_file/ac860a09-78e2-4eb5-b8fb-0aea8b8d8a8e/ScienceOpen/022_Morris.pdf
https://ds.amu.edu.et/xmlui/bitstream/handle/123456789/17461/Graham%20C.%20Goodwin%2C%20Stefan%20F.%20Graebe%2C%20Mario%20E.%20Salgado-Control%20System%20Design%20-Prentice%20Hall%20%282000%29.pdf?sequence=1&isAllowed=y
https://link.springer.com/book/10.1007/1-84628-493-7
http://kundoku.free.fr/O/008/Optimal%20Sampled%20Data%20Control%20Systems/Optimal%20Sampled%20Data%20Control%20Systems.pdf
https://books.google.com/books?hl=en&lr=&id=vcc_AAAAQBAJ&oi=fnd&pg=PR27&dq=Industrial+Control+System&ots=iwvsVisLQs&sig=lyE7Iv7kcU94C9pCnQRd7XMzvCM
https://www.academia.edu/download/30480284/a_survey_on_analysis_and_design_of_model-based_fuzzy_control_systems-dsk.pdf
https://core.ac.uk/download/pdf/19878056.pdf
https://www.academia.edu/download/39283936/CENG_120_Textbook.pdf
https://eprints.gla.ac.uk/3817/1/IEEE3.pdf?re

In [8]:
i=0
response_i=requests.get(item.select('a')[0]['href'])
for item in soup.select('[data-lid]'):
  if ('.pdf' in item.select('a')[0]['href'] ):
     i+=1

     print(item.select('a')[0]['href'])
     pdf = open("pdf"+str(i)+".pdf", 'wb')
     pdf.write(response_i.content)
     pdf.close()
     print("File ", i, " downloaded")
print(i)


http://www.gocs.com.de/pages/fachberichte/archiv/164-sp800_82_r2_draft.pdf
File  1  downloaded
https://www.scienceopen.com/document_file/ac860a09-78e2-4eb5-b8fb-0aea8b8d8a8e/ScienceOpen/022_Morris.pdf
File  2  downloaded
https://ds.amu.edu.et/xmlui/bitstream/handle/123456789/17461/Graham%20C.%20Goodwin%2C%20Stefan%20F.%20Graebe%2C%20Mario%20E.%20Salgado-Control%20System%20Design%20-Prentice%20Hall%20%282000%29.pdf?sequence=1&isAllowed=y
File  3  downloaded
http://kundoku.free.fr/O/008/Optimal%20Sampled%20Data%20Control%20Systems/Optimal%20Sampled%20Data%20Control%20Systems.pdf
File  4  downloaded
https://www.academia.edu/download/30480284/a_survey_on_analysis_and_design_of_model-based_fuzzy_control_systems-dsk.pdf
File  5  downloaded
https://core.ac.uk/download/pdf/19878056.pdf
File  6  downloaded
https://www.academia.edu/download/39283936/CENG_120_Textbook.pdf
File  7  downloaded
https://eprints.gla.ac.uk/3817/1/IEEE3.pdf?ref=driverlayer.com
File  8  downloaded
8


# Extracting Sentences from the downloaded pdf files

To extract text from pdf, we will use PyPDF2 library.

In [10]:
pip install PyPDF2

In [11]:
from PyPDF2 import PdfReader,PdfFileWriter

In [14]:
reader_1=PdfReader('Guide to Industrial Control System Security.pdf')
reader_2=PdfReader('Cyber Attacks ICS.pdf')
reader_3=PdfReader('IEEE3.pdf')
reader_4=PdfReader('Optimal Sampled Data Control Systems.pdf')

In [15]:
print(len(reader_1.pages))  #number of pages in the pdf
print(len(reader_2.pages))
print(len(reader_3.pages))
print(len(reader_4.pages))

255
8
19
420


In [16]:
def extract_text(reader):
  i=1
  for i in range(len(reader.pages)):
    page=reader.pages[i]
    i+=1
    text=page.extract_text()
    print('Page',i)
    print(text)
    print('       ')

In [58]:
text_1=extract_text(reader_1)
text_2=extract_text(reader_2)
text_3=extract_text(reader_3)
#text_4=extract_text(reader_4)

Streaming output truncated to the last 5000 lines.
SA-17 Developer Security Architecture and Design  Not Selected  Not Selected  SA-17 
SC-1 System and Communications Protection 
Policy and Procedures  SC-1  SC-1  SC-1  
SC-2 Application Partitioning  Not Selected  SC-2  SC-2  
SC-3 Security Function Isolation  Not Selected  Not Selected  SC-3 
SC-4 Information in Shared Resources  Not Selected  SC-4  SC-4  
SC-5 Denial of Service Protection  SC-5  SC-5  SC-5  
SC-7 Boundary Protection  SC-7  SC-7 (3) (4) (5) 
(7) (18) SC-7 (3) (4) (5) 
(7) (8)  (18)  
(21) 
SC-8 Transmission Confidentiality and Integrity  Not Selected  SC-8 (1) SC-8 (1) 
SC-10 Network Disconnect  Not Selected  SC-10  SC-10  
SC-12 Cryptographic Key Establishment and 
Management   SC-12 SC-12  SC-12 (1)  
SC-13 Cryptograph ic Protection  SC-13 SC-13  SC-13  
SC-15 Collaborative Computing Devices  SC-15 SC-15 SC-15  
SC-17 Public Key Infrastructure Certificates  Not Selected  SC-17 SC-17  
SC-18 Mobile Code  Not Selecte

Above was the text from the pdf.Now we will perform tokenization to extract sentences. For this, there are many libraries . I have used NLTK library which can be installed using pip.

In [20]:
!pip install nltk

In [21]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [22]:
from nltk.tokenize import sent_tokenize

PDF1

Extracting sentences from each page of the pdf. I have started from page 10 since the first 9 pages of the pdf had nothing but author names.

In [38]:
i=10
text=" "
for i in range(len(reader_1.pages)):
  page=reader_1.pages[i]
  text += page.extract_text()
  i+=1
  sentences1=sent_tokenize(text)
print(sentences1)
print(sentences1[400])
print(len(sentences1[400].split()))
print('.................')


[' NIST Special Publication 800 -82 \nRevision 2 Initial Public Draft  \n \n \n \nGuide to Industrial Control \nSystems (ICS) Security  \nSupervisory Control and Data Acquis ition (SCADA) Systems, Distributed Control Systems (DCS), \nand Other Control System Configurations suc h as Programmable Logic Controllers (PLC)  \n \nKeith Stouffer  \nSuzanne Lightman  \nVictoria Pillitteri  \nMarshall Abrams  \nAdam Hahn  \n \n \n  \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n \n   NIST Special Publication 800 -82 \nRevision 2 Initial Public Draft  \n \n \nGuide to Industrial Control Systems \n(ICS) Secur ity \nSupervisory Control and Data Acquisition (SCADA) systems, Distributed Control Systems (DCS), and \nother control system configurations such as Programmable Logic Controllers (PLC)  \n \n \nKeith Stouffer  \nIntelligent Systems Division   \nEngineering Laboratory  \n \nSuzanne Lightman  \nVictoria Pillitteri  \nComputer Security Division   \nInformation Technology Laboratory   \n \nMars

len(reader.pages) gives the number of pages in the pdf. sent_tokenize is  used to convert sentences. By sentences we mean the strings that start and end with a dot.By using.split() method, I have determined the number of words in the sentence at index 400(in this case) in the list named sentences.

In [39]:
long_sentences1=[]
j=0
for j in range(len(sentences1)):
  if len(sentences1[j].split())>=10:
    long_sentences1.append(sentences1[j])
  j+=1
print(len(long_sentences1))


3654


PDF 2

In [30]:
i=1
text=" "
for i in range(len(reader_2.pages)):
  page=reader_2.pages[i]
  text += page.extract_text()
  i+=1
  sentences_2=sent_tokenize(text)
print(sentences_2)
print(sentences_2[200])
print(len(sentences_2[200].split()))
print('.................')

[' © Morris et al.', 'Published by BCS  \nLearning and Development Ltd.', 'Proceedings of the 1st International Symposiu m for ICS & SCADA Cyber Security Research 2013  \n22 Industrial Control System Cyber Attacks \nThomas H. Morris1, Wei Gao \nMississippi State University \nMississippi State, MS, USA \n1morris@ece.msstate.edu \nThis paper presents a set of attacks against SCAD A control systems.', 'The attacks are grouped into \n4 classes; reconnaissance, response and measurement injection,  command injection and denial of \nservice.', 'The 4 classes are defined and each attack is described in detail.', 'The response and \nmeasurement injection and command injection cl asses are subdivided into  sub-classes based on \nattack complexity.', 'Each attack described in th is paper has been exercised against industrial \ncontrol systems in a laboratory setting.', 'Industrial Control System.', 'Threat Model.', 'Taxonomy.', '1.', 'INTRODUCTION \nSupervisory Control and Data Acquisition (SCADA

In [41]:
long_sentences2=[]
j=0
for j in range(len(sentences_2)):
  if len(sentences_2[j].split())>=10:
    long_sentences2.append(sentences_2[j])
  j+=1
print(len(long_sentences2))


254


PDF 3

In [45]:
i=1
text=" "
for i in range(len(reader_3.pages)):
  page=reader_3.pages[i]
  text += page.extract_text()
  i+=1
  sentences_3=sent_tokenize(text)
print(sentences_3)
print(sentences_3[200])
print(len(sentences_3[200].split()))
print('.................')

['  \n \n \n \n \n \nAng, K.H.', 'and Chong, G.C.Y.', 'and Li, Y.', '(2005 ) PID control system  \nanalysis, design, and technology.', 'IEEE Transactions on Control Systems \nTechnology  13(4):pp.', '559-576 .', 'http:/ /eprints.gla.ac.uk/3817/ \n \nDeposited on: 13 November 2007 \n \n \nGlasgow ePrints Service \nhttp://eprints.gla.ac.uk IEEE TRANSACTIONS ON CONTROL SYSTEMS TECHNOLOGY , VOL.', '13, NO.', '4, JULY 2005 559\nPID Control System Analysis, Design,\nand Technology\nKiam Heong Ang, Gregory Chong , Student Member, IEEE , and Yun Li , Member, IEEE\nAbstract— Designing and tuning a proportional-integral-deriva-\ntive (PID) controller appears to be conceptually intuitive, but canbe hard in practice, if multiple (and often conflicting) objectivessuch as short transient and high stability are to be achieved.Usually, initial designs obtained by all means need to be adjustedrepeatedly through computer simulations until the closed-loopsystem performs or compromises as desired.', 'This 

In [47]:
long_sentences3=[]
j=0
for j in range(len(sentences_3)):
  if len(sentences_3[j].split())>=10:
    long_sentences3.append(sentences_3[j])
  j+=1
print(len(long_sentences3))


353


PDF 4

In [56]:
i=1
text=" "
for i in range(len(reader_4.pages)):
  page=reader_4.pages[i]
  text += page.extract_text()
  i+=1
  sentences_4=sent_tokenize(text)
print(sentences_4)
print(sentences_4[0])
print(len(sentences_4[0].split()))
print('.................')

[' /C7/D4/D8/CX/D1/CP/D0 /CB/CP/D1/D4/D0/CT/CS/B9/BW/CP/D8/CP /BV/D3/D2 /D8/D6/D3/D0/CB/DD/D7/D8/CT/D1/D7/CC /D3/D2/CV/DB /CT/D2 /BV/CW/CT/D2 /CP/D2/CS /BU/D6/D9/CR/CT /BY /D6/CP/D2/CR/CX/D7/CE /CT/D6/D7/CX/D3/D2 /BG/BA/BG/B8 /BW/CT/CR/CT/D1 /CQ/CT /D6 /BD/BL/BL/BG/CC /D3 /C5/CX/D2/CV /CP/D2/CS /C2/CX/D2/CV/DB /CT/D2/B8 /C2/CT/D7/D7/CX/CT /CP/D2/CS /C4/CX/CP/D2/DA/CX/CX/CX/C8/D6/CT/CU/CP/CR/CT/C5/CP/D2 /DD /D8/CT/CR /CW/D2/CX/D5/D9/CT/D7 /CP/D6/CT /CP /DA /CP/CX/D0/CP/CQ/D0/CT /CU/D3/D6 /CS/CT/D7/CX/CV/D2/CX/D2/CV /D0/CX/D2/CT/CP/D6 /D1 /D9/D0/D8/CX/DA /CP/D6/CX/CP/CQ/D0/CT /CP/D2/CP/B9/D0/D3/CV /CR/D3/D2 /D8/D6/D3/D0/D0/CT/D6/D7/BM /D4 /D3/D0/CT /D4/D0/CP/CR/CT/D1/CT/D2 /D8/D9 /D7/CX/D2/CV /D3/CQ/D7/CT/D6/DA /CT/D6/B9/CQ/CP/D7/CT/CS /CR/D3/D2 /D8/D6/D3/D0/D0/CT/D6/D7/B8 /D0/D3 /D3/D4/B9/D7/CW/CP/D4/CX/D2/CV/B8 /D8/CW/CT /CX/D2 /DA /CT/D6/D7/CT /C6/DD/D5/D9/CX/D7/D8 /CP/D6/D6/CP /DD /D1/CT/D8/CW/D3 /CS/B8 /CR/D3/D2 /DA /CT/DC /D3/D4/D8/CX/D1/CX/DE/CP/D8/CX/D3/D2 /CX/D2/CR/D3/D2 /D8/D6/

In [57]:
long_sentences4=[]
j=0
for j in range(len(sentences_4)):
  if len(sentences_4[j].split())>=10:
    long_sentences4.append(sentences_4[j])
  j+=1
print(len(long_sentences4))

1


# Writing the sentences in a text file

We have created a list named Long_sentences which contains those sentences which contain more then 10 words. The number of such sentences is 3654 for this pdf. Now, we will write the sentences into a text file.

In [43]:
file_path="pdf1.txt"
#opening the file in write mode
with open(file_path,'w',encoding='utf-8')as file:
  #writing each element in the list to a new line
  for element in long_sentences1:
    file.write(element + '\n')
print("The list has been written to filepath:",file_path)

The list has been written to filepath: pdf1.txt


In [49]:
file_path="pdf2.txt"
#opening the file in write mode
with open(file_path,'w',encoding='utf-8')as file:
  #writing each element in the list to a new line
  for element in long_sentences2:
    file.write(element + '\n')
print("The list has been written to filepath:",file_path)

The list has been written to filepath: pdf2.txt


In [51]:
file_path="pdf3.txt"
#opening the file in write mode
with open(file_path,'w',encoding='utf-8')as file:
  #writing each element in the list to a new line
  for element in long_sentences3:
    file.write(element + '\n')
print("The list has been written to filepath:",file_path)

The list has been written to filepath: pdf3.txt
